### ISSUE TITLE
Non-matmul Kernel factories don't actually compile their op (structural)

### ISSUE THREAD

**Issue by Ravenwater** [stillwater-sc/kpu-sim#18]
## TL;DR
\`KernelCompiler\` only knows how to compile matmul. Every \`Kernel::create_*\` factory for non-matmul ops (softmax, elementwise, pool2d, conv2d, layernorm, batchnorm, reduce) calls \`compile_matmul\` to get a Kernel framework, then stamps a different \`op_type_\` over it without changing the underlying instruction stream. The runtime path (\`executor.execute(kernel.program())\`) runs whatever is in \`program_\`, so launching a "softmax kernel" today doesn't compute a softmax — it computes a matmul.

PR #17 made the matmul tiny (1x1x1) so it stops OOMing on Windows for big shapes, but does **not** fix correctness.

## Surface
All non-matmul kernel factories in \`src/system/simulator/kernel.cpp\` are affected:

| Factory | Placeholder | Severity |
|---|---|---|
| \`create_conv2d\` (line 182) | \`compile_matmul(M, N, K, opts)\` — at least shape-aware, possibly an intentional im2col? | Unclear: needs design clarification |
| \`create_layernorm\` (line 239) | \`compile_matmul(1, D, D, opts)\` | Wrong op |
| \`create_batchnorm\` (line 288) | \`compile_matmul(1, D, D, opts)\` | Wrong op |
| \`create_reduce\` (line 328) | \`compile_matmul(1, C, C, opts)\` | Wrong op |
| \`create_elementwise\` (line 371) | \`compile_matmul(1, 1, 1, opts)\` (post-#17) | Wrong op |
| \`create_pool2d\` (line 440) | \`compile_matmul(1, 1, 1, opts)\` (post-#17) | Wrong op |
| \`create_softmax\` (line 534) | \`compile_matmul(1, 1, 1, opts)\` (post-#17) | Wrong op |

Only \`create_matmul\` (line 54) produces a faithful program.

## Why kernel_test still passes
\`tests/driver/test_kernel.cpp\` only checks metadata (\`kernel.softmax_config()\`, \`kernel.total_flops()\`, shape introspection) — it never executes the kernel via \`executor.execute(kernel.program())\`. So the structural bug is invisible to that test, even after PR #17 took the OOM away.

The downstream paths that actually \`execute()\` a kernel are:
- \`src/software/runtime/runtime.cpp:164\` — \`KPURuntime::launch\`
- \`src/tools/benchmark/benchmark.cpp:29,135,164,195\` — the perf benchmarks

If those benchmarks happen to produce numeric output for non-matmul ops, the numbers are currently wrong. Worth confirming with whoever owns the benchmark suite.

## Right shape of fix
Add real compile methods to \`KernelCompiler\` (\`include/sw/compiler/kernel_compiler.hpp\`), one per op type that today fakes it:

\`\`\`cpp
Kernel compile_softmax(const SoftmaxConfig& cfg, const CompileOptions& opts);
Kernel compile_elementwise(const ElementwiseConfig& cfg, const CompileOptions& opts);
Kernel compile_pool2d(const Pool2DConfig& cfg, const CompileOptions& opts);
Kernel compile_conv2d(const Conv2DConfig& cfg, const CompileOptions& opts);
Kernel compile_layernorm(const LayerNormConfig& cfg, const CompileOptions& opts);
Kernel compile_batchnorm(const BatchNormConfig& cfg, const CompileOptions& opts);
Kernel compile_reduce(const ReduceConfig& cfg, const CompileOptions& opts);
\`\`\`

Each one should:
1. Produce an instruction stream the executor can actually run for that op type (probably new opcodes in the DM ISA, or a per-op emitter built on top of the existing dataflow primitives).
2. Replace the matmul-piggyback path in the corresponding \`create_*\` factory.

Open design questions worth a brief huddle before coding:
- Should the runtime instead dispatch on \`op_type\` and route to op-specific executors, leaving \`program_\` op-neutral? That's the alternative architecture — fewer instruction-stream variants, more executor surface.
- Whichever direction, what's the validation strategy? \`kernel_test\` currently only checks metadata; we'd need executor-level numerical tests against a reference implementation (numpy or hand-rolled) for each op.

## Acceptance criteria
- [ ] \`KernelCompiler\` has at least one real (non-matmul-piggyback) compile method for each currently-faked op type.
- [ ] All \`Kernel::create_*\` factories route to the matching \`compile_*\`.
- [ ] At least one executor-level numerical test exists for each formerly-faked op type, verifying output against a reference (numpy / hand-rolled).
- [ ] Benchmark suite (\`src/tools/benchmark/benchmark.cpp\`) is verified to produce correct numbers for the non-matmul ops, or its mention of those ops is documented as performance-only-no-correctness.
- [ ] Comments / docstrings in \`kernel.cpp\` no longer describe matmul placeholders.

## Related
- Surfaced (and partially masked) by #17, which only shrank the placeholder matmul so Windows MSVC stopped OOMing on the GPT-2-shaped softmax test in \`kernel_test\`.
- Same shape of root cause as #11 / #13 (broken API patched in tests rather than fixed): an "it compiles, the test for it asserts on metadata not behaviour" smell.

## Out of scope
- The deeper question of "should the runtime dispatch on op_type instead of executing program_ blindly?" — that's an architectural design call, worth its own discussion before committing to a direction.

**Comment by Ravenwater** [comment]
The structural half is fixed in PR (linked above): all non-matmul factories now compile op-faithful loop-based streaming programs (no matmul piggyback), conv2d is documented as intentional im2col+GEMM (answering this issue's open question), and ConcurrentExecutor interprets the loop/AUTO/VE instruction forms.

Remaining for full closure - the numerical half: VE_ELEMENTWISE/VE_REDUCE functional semantics + executor-level oracle tests per op. That is the capability-closure work of the pattern epics E2 (#71, broadcast/elementwise) and E3 (#72, online reductions), where the VE operand-encoding design belongs. This issue stays open until those land.

**Comment by Ravenwater** [comment]
Reopening: the merge commit's 'resolves the structural half of #18' was parsed as a closing keyword. The structural half IS fixed (PR #142); this issue tracks the remaining numerical half - VE_ELEMENTWISE/VE_REDUCE functional semantics + executor-level oracle tests, landing with pattern epics E2 (#71) and E3 (#72).

**Comment by Ravenwater** [comment]
The elementwise half of this issue is complete with E2 (#71): VE_ELEMENTWISE has full numerical semantics (#100), value-producing CSP execution verified against host oracles across all 14 ops/forms (#102), and a form x size x envelope regression matrix with credit invariants (#103, PR #148). The remaining half - VE_REDUCE numerical semantics - belongs to the online-reduce epic E3 (#72).

**Comment by Ravenwater** [comment]
Closed by epic E3 (#72). The numerical half of the VE ops is complete on both value-producing tiers:
- **VE_ELEMENTWISE**: full semantics via VEOperands (#100), value-producing CSP execution vs host oracles (#102), regression matrix (#103) - epic E2 (#71).
- **VE_REDUCE**: VEReduceOperands + behavioral running-combine (#105), value-producing CSP reduction vs host oracles (#107), regression matrix (#108) - epic E3 (#72).

Both the behavioral ISA executor and the CSP functional tier compute correct values for every VE op, verified against independent host oracles.
